rs/https_outcalls/consensus/src/payload_builder.rs

```
@@ -282,16 +289,32 @@ impl CanisterHttpPayloadBuilderImpl {
 
             let candidates_and_divergences = response_candidates_by_callback_id
                 .into_iter()
-                .filter_map(|(_, grouped_shares)| {
-                    if let Some((metadata, shares)) = grouped_shares.iter().find(|(_, shares)| {
-                        unique_responses_count += 1;
-                        let signers: BTreeSet<_> =
-                            shares.iter().map(|share| share.signature.signer).collect();
-                        // We need at least threshold different signers to include the response
-                        signers.len() >= threshold
-                    }) {
-                        // A set of grouped shares large enough to meet the
-                        // threshold was found, we should produce a result.
+                .filter_map(|(id, grouped_shares)| {
+                    let consensus_candidate =
+                        grouped_shares.iter().find_map(|(metadata, shares)| {
+                            unique_responses_count += 1;
+
+                            if let Some(context) = canister_http_request_contexts.get(&id) {
+                                if let Replication::NonReplicated(node_id) = context.replication {
+                                    // For a non-replicated call, we require EXACTLY ONE share,
```

where do we check that we have *exactly* one share? as far as I can tell we only check that there is at least one share from the designated node. A malicious designated node could produce two valid responses to the same request, do we handle it somehow?

---

Hm good point. I guess if the designated node is malicious, then it can anyway send whatever shares it likes. (note that if the block maker is not designated, then it doesn't have the content, so it can't add it to the payload).
We somewhat assume that if the designated node is malicious, the respective request can anyway be non legitimate. But still we should handle it gracefully, so maybe this is a question fro execution?

@dfinity/execution what would happen if two responses for the same request were send in the same batch?

---

> We somewhat assume that if the designated node is malicious, the respective request can anyway be non legitimate

That's true, in general we can't know if the content of a response is non-legitimate, but I think we should be able to argue that only a malicious block maker would include two responses for the same request and therefore the block could be thrown away, even if it's harmless. Especially since we already [have a check](https://sourcegraph.com/github.com/dfinity/ic@19502f0c93b5271fa9ce54f17426bbf47ae22df3/-/blob/rs/https_outcalls/consensus/src/payload_builder.rs?L460-465) that a response wasn't submitted already

---

Regarding "what would happen if we return the same message twice?": IIUC the second message would be [silently ignored](https://sourcegraph.com/github.com/dfinity/ic/-/blob/rs/execution_environment/src/execution_environment.rs?L523) as no call context can be found after executing the first one.

I agree with Kamil, though, that it would be good practice to check for duplicates, as long as we make sure that we have the checks in both the payload builder and the validator.

---

I think we have to distinguish here: from the viewpoint of Consensus, sure, payloads should just look like a blob of bytes and Consensus should not have to look into them. However, inside the implementation of a payload builder (which is not part of Consensus only the payload builder trait is) it is fine to know about the details and also do all kinds of specific checks. 

As per my understanding this is precisely how this abstraction via the payload builder trait was intended. For example, in XNet we build the payload taking lots of details into account which are important for XNet payloads; however Conensus should not be aware of these details and only view the XNet payload as abstract blob of bytes. Similarly in ingress payload building.

Based on the interpretation above it would seem that the suggested check does not break any abstractions, no?

---

XNet is the odd one out here. It's the only one that enforces constraints between it's messages.
The other payload builders just check some basic constraints on the payload itself, such as size or maximum number of messages and then validate the messages individually.

I was hoping that one day we would be able to move away from this interface and do away with the payload-types and see the payload as just an array of tagged messages, with a per message-type validator. This would be a simpler interface and allow for much more flexible payload building.

That said, given that I don't have an idea how to get there with respect to XNet and I don't think that this single check here makes a huge difference and would block us in the future, it shouldn't slow us down. I am fine with it either way 👍

---

Ah I see. So you are thinking about an interface where you can just say "give me the next message" as opposed to "give me the entire payload". Makes sense.

---

IIUC since we're checking against the validation context only, and since the validation context may not move between two subsequent blocks, it's possible that two subsequent leaders include the response to the same request anyways? Which is to say, we have to rely on the execution ignoring duplicate responses anyways?